### PR TITLE
[FLINK-11164] Check for sentinel values when creating new Kinesis ShardIterator

### PR DESCRIPTION
**Pull request by aljoscha**
@tzulitai Would this be a fix for the issue?

### CODE REVIEW

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java [hunk 1]
```diff
@@ -380,7 +381,19 @@ private GetRecordsResult getRecords(String shardItr, int maxNumberOfRecords) thr
 			} catch (ExpiredIteratorException eiEx) {
 				LOG.warn("Encountered an unexpected expired iterator {} for shard {};" +
 					" refreshing the iterator ...", shardItr, subscribedShard);
-				shardItr = kinesis.getShardIterator(subscribedShard, ShardIteratorType.AFTER_SEQUENCE_NUMBER.toString(), lastSequenceNum.getSequenceNumber());
+				if (SentinelSequenceNumber.getAllSentinelSequenceNumbers().contains(lastSequenceNum)) {
+					// we cannot create a shard iterator because we still have an initial sentinel
+					// value as the current sequence number
+					LOG.warn(
```
[the remark this answers is not in the document]
Comment (aljoscha) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java]: Could we just call `getInitialShardIterator()`? If not, I'll probably factor out the if-branches that deal with sentinel values into a new method. WDYT?

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/model/SentinelSequenceNumber.java [hunk 2]
```diff
@@ -54,4 +57,17 @@
 	public SequenceNumber get() {
 		return sentinel;
 	}
+
+	/**
+	 * Returns a {@link List} of {@link SequenceNumber SequenceNumbers} that are used as sentinel
+	 * values.
+	 */
+	public static List<SequenceNumber> getAllSentinelSequenceNumbers() {
+		List<SequenceNumber> result = new ArrayList<>(values().length);
```
[the remark this answers is not in the document]
Comment (aljoscha) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/model/SentinelSequenceNumber.java]: Makes sense, yes.

**Comment by aljoscha** [comment]
@tzulitai I refactored it a bit

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java [hunk 3]
```diff
@@ -149,52 +150,72 @@ protected String getInitialShardIterator() throws Exception {
 
 		// before infinitely looping, we set the initial nextShardItr appropriately
 
-		if (lastSequenceNum.equals(SentinelSequenceNumber.SENTINEL_LATEST_SEQUENCE_NUM.get())) {
+		if (isSentinelSequenceNumber(lastSequenceNum)) {
+			nextShardItr = getShardIteratorForSentinel(lastSequenceNum);
+		} else {
+			// we will be starting from an actual sequence number (due to restore from failure).
+			// if the last sequence number refers to an aggregated record, we need to clean up any dangling sub-records
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java]: comments here about record aggregation should go into the `getShardIteratorForRealSequenceNumber` method

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java [hunk 4]
```diff
@@ -149,52 +150,72 @@ protected String getInitialShardIterator() throws Exception {
...
+		return nextShardItr;
+	}
 
-				// get only the last aggregated record
-				GetRecordsResult getRecordsResult = getRecords(itrForLastAggregatedRecord, 1);
+	protected String getShardIteratorForRealSequenceNumber(SequenceNumber sequenceNumber) throws Exception {
+		String nextShardItr;
 
-				List<UserRecord> fetchedRecords = deaggregateRecords(
+		if (sequenceNumber.isAggregated()) {
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java]: While you're at refactoring this:
Can we have the logic in the `sequenceNumber.isAggregated()` branch in a separate method?
Or
invert the if-else branch.

Just for better code readability.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java [hunk 5]
```diff
@@ -149,52 +150,72 @@ protected String getInitialShardIterator() throws Exception {
...
+		if (isSentinelSequenceNumber(lastSequenceNum)) {
+			nextShardItr = getShardIteratorForSentinel(lastSequenceNum);
+		} else {
+			// we will be starting from an actual sequence number (due to restore from failure).
+			// if the last sequence number refers to an aggregated record, we need to clean up any dangling sub-records
+			// from the last aggregated record; otherwise, we can simply start iterating from the record right after.
+
+			nextShardItr = getShardIteratorForRealSequenceNumber(lastSequenceNum);
+		}
+		return nextShardItr;
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java]: `nextShardItr` is a redundant local variable now.

==>

```
protected String getInitialShardIterator() throws Exception {
    if (!isSentinelSequenceNumber(lastSequenceNum)) {
        return getShardIteratorForRealSequenceNumber(lastSequenceNum);
    }
    return getShardIteratorForSentinel(lastSequenceNum);
}
```

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java [hunk 6]
```diff
@@ -381,13 +402,15 @@ private GetRecordsResult getRecords(String shardItr, int maxNumberOfRecords) thr
 			} catch (ExpiredIteratorException eiEx) {
 				LOG.warn("Encountered an unexpected expired iterator {} for shard {};" +
 					" refreshing the iterator ...", shardItr, subscribedShard);
-				if (SentinelSequenceNumber.getAllSentinelSequenceNumbers().contains(lastSequenceNum)) {
+				if (isSentinelSequenceNumber(lastSequenceNum)) {
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java]: Now this if-else branch is basically identical to the `getInitialShardIterator` method, minus the fact that `lastSequenceNum` will never be in the middle of an aggregated record (which is fine).
I don't think the new warning log adds too much value, TBH.

So, maybe rename `getInitialShardIterator` to just `getShardIterator` and use it here.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java [hunk 7]
```diff
@@ -149,52 +150,72 @@ protected String getInitialShardIterator() throws Exception {
...
-			if (lastSequenceNum.isAggregated()) {
-				String itrForLastAggregatedRecord =
-					kinesis.getShardIterator(subscribedShard, ShardIteratorType.AT_SEQUENCE_NUMBER.toString(), lastSequenceNum.getSequenceNumber());
+		return nextShardItr;
+	}
 
-				// get only the last aggregated record
-				GetRecordsResult getRecordsResult = getRecords(itrForLastAggregatedRecord, 1);
+	protected String getShardIteratorForRealSequenceNumber(SequenceNumber sequenceNumber) throws Exception {
+		String nextShardItr;
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java]: Probably can get rid of this local variable, as suggested above.

**Comment by aljoscha** [comment]
I refactored this some more. PTAL. 😄

**Comment by tzulitai** [comment]
@aljoscha LGTM 👍

**Pull request closed without merging**